src/vsr/client.zig

```diff
@@ -373,19 +579,90 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 on_reply_callback(self, inflight.message, reply);
             }
 
-            if (inflight.callback) |callback| {
+            if (inflight.demux_list) |demux_list| {
                 assert(inflight.message.header.operation != .register);
 
-                callback(
-                    inflight.user_data,
-                    inflight.message.header.operation.cast(StateMachine),
-                    reply.body(),
-                );
+                const operation = inflight.message.header.operation.cast(StateMachine);
+                switch (operation) {
+                    .reserved, .root, .register => unreachable,
+                    .create_accounts => self.decode_reply(.create_accounts, demux_list, reply),
+                    .create_transfers => self.decode_reply(.create_transfers, demux_list, reply),
+                    .lookup_accounts => self.decode_reply(.lookup_accounts, demux_list, reply),
+                    .lookup_transfers => self.decode_reply(.lookup_transfers, demux_list, reply),
+                }
             } else {
                 assert(inflight.message.header.operation == .register);
             }
         }
 
+        /// Decoding the state machine reply into multiple logical batches,
+        /// Example for create_accounts and create_transfers:
+        ///
+        /// Response (index,result):
+        /// (0,x),(1,x),(3,x),(5,x),(20,x)
+        ///
+        /// Demux list (offset,size):
+        /// (0,2)(2,10),(12,8),(20,1)
+        ///
+        /// Demux process:
+        /// Reply        | Demux  | Demuxed reply
+        /// -------------|--------|---------------
+        /// (0,x),(1,x)  | (0,2)  | (0,x),(1,x)
+        /// (3,x),(5,x)  | (2,10) | (1,x),(3,x)
+        /// ()           | (12,8) | ()
+        /// (20,x)       | (20,1) | (0,x)
+        fn decode_reply(self: *Self, comptime operation: StateMachine.Operation, demux_list: DemuxList, reply: *Message) void {
```

Comment:
This logic will (in part) need to move into `StateMachine`, since the `result.index` mapping is statemachine-specific logic.